### PR TITLE
test: skip terminfo parsing in Miri

### PR DESCRIPTION
**Pull request by RalfJung**
Terminfo parsing takes a significant amount of time in Miri, making libtest startup very slow. To work around that Miri in fact unsets the `TERM` variable. However, this means we don't get colors in `cargo miri test`.

So I propose we add some logic in libtest that skips parsing terminfo files under Miri, and just uses the regular basic coloring commands (taken from the `colored` crate).

As far as I can see, these two commands are all that libtest ever needs from terminfo, so Miri doesn't even lose any functionality through this. If you want I can entirely remove the terminfo parsing code and just use these commands instead.

Cc https://github.com/rust-lang/miri/issues/2292 @saethlin

### CODE REVIEW

File: library/test/src/term/terminfo/mod.rs [hunk 1]
```diff
@@ -119,13 +130,22 @@ pub(crate) struct TerminfoTerminal<T> {
 impl<T: Write + Send> Terminal for TerminfoTerminal<T> {
     fn fg(&mut self, color: color::Color) -> io::Result<bool> {
         let color = self.dim_if_necessary(color);
+        if cfg!(miri) && color < 8 {
```
Comment (saethlin) [inline library/test/src/term/terminfo/mod.rs:133]: It's not obvious to me what happens if `color = 9`. The logic for `num_colors` isn't totally straightforward, but I think in Miri we end up with `num_colors = 0`, right? Could we explain this in a comment?

**Comment by saethlin** [comment]
Wow this is a lot less code than I expected!

File: library/test/src/term/terminfo/mod.rs (hunk 1 above)
Comment (RalfJung) [reply library/test/src/term/terminfo/mod.rs:133]: Yeah `num_colors` is always 0 in Miri. I added a comment.

File: library/test/src/term/terminfo/mod.rs [hunk 2]
```diff
@@ -119,13 +130,24 @@ pub(crate) struct TerminfoTerminal<T> {
...
+            // The Miri logic for this only works for the most basic 8 colors, which we just assume
+            // the terminal will support. (`num_colors` is always 0 in Miri, so higher colors will
+            // just fail. But libtest doesn't use any higher colors anyway.)
+            let s = format!("\x1B[3{color}m");
```
Comment (thomcc) [inline library/test/src/term/terminfo/mod.rs:137]: Huh, slightly surprised this is more efficient under miri than something like `write!(self.out, "\x1B[3{color}m")`.

Comment (RalfJung) [reply library/test/src/term/terminfo/mod.rs:137]: I don't think it is more efficient. But Miri isn't *that* slow, a bit of formatting is fine.

The terminfo parser is doing a *lot* of work and constructing a huge data structure to represent the parsed result, that's why this is noticeable.

But your code looks nicer anyway. :D  So, happy to change it accordingly.

**Comment by thomcc** [comment]
@bors r+

(P.S. Let me know if you need review on any unix terminal or windows console stuff inside miri, terminals are a bit of an obsession of mine, so I'm pretty familiar with how this all works. I'm guessing this is probably enough for now though)

**Comment by RalfJung** [comment]
@bors r-

I'll fix that nit you mentioned

**Comment by RalfJung** [comment]
> (P.S. Let me know if you need review on any unix terminal or windows console stuff inside miri, terminals are a bit of an obsession of mine, so I'm pretty familiar with how this all works. I'm guessing this is probably enough for now though)

Thanks! Miri does not really do anything with the terminal, it just forwards the bytes it gets from the interpreted program. Though https://github.com/rust-lang/rust/issues/98070 will be nice for Miri since it means we can implement the `isatty` shim in a better way. :)

However you might then be the right person for the other question I had in this PR: given that, from what I can see, these two command codes are literally everything libtest ever uses -- do you think it would make sense to remove all that terminfo parsing and just use these commands? (E.g. the `colored` crate doesn't do any terminfo parsing and seems to work fine, though maybe it just hasn't been tested enough.) In particular I don't know if that will work on Windows.

**Comment by thomcc** [comment]
> do you think it would make sense to remove all that terminfo parsing and just use these commands? (E.g. the colored crate doesn't do any terminfo parsing and seems to work fine, though maybe it just hasn't been tested enough.) 

I'd have to look what all it uses. If it's just 8 colors and reset, this uses a very consistent set of escape sequences across a wide range of terminals -- you need to start worrying about terminfo mostly when you care about advanced features, especially events handling, since the sequences terminals send for events are often very different, even among modern terminals which in most other ways have converged on doing things like more or less xterm.

That said, removing this would probably break compatibility with certain retro terminals though (which currently we probably support... But really, those terminals should just disabling color anyway, I suspect.

If we're only interested in basic coloring, removing this is probably a good idea, as it will speed things up for everybody to not query the terminfo DB. It will also avoid doing weird things if users have `$TERM` set to a goofy value (like `TERM=vt100`, which I've seen before).

That said, I'd like to look through terminfo.src to see if there's anything people are likely to still use that have other values for these (I know often reset (sgr0) will often reset the charset too, but we don't change this, so it's fine).

> In particular I don't know if that will work on Windows.

Windows doesn't have a terminfo db, and uses the console apis (e.g. winapi functions which probably turn into syscalls) to change the color. I believe we already do this in libtest, so windows isn't a concern.

On modern windows (Win10 and later) the escape sequences are supported, although you may need to explicitly ask the console to `ENABLE_VIRTUAL_TERMINAL_PROCESSING`. This is actually on by default on many consoles, but has the benefit of returning an error on older windows and if the terminal can't support it.

That said, this is fiddley and for simple stuff like we want, so I wouldn't bother changing away from the console apis (besides, we still support win7 and win8 at the moment).

**Pull request merged**